### PR TITLE
CI unpin flake8, only run flake8-rst in pre-commit

### LINKED ISSUE #34150: CI: Linting errors from flake >= 3.8.1
The new version of flake bumps the version of pycodestyle, which results in several linting errors on CI.  These include:

- (many instances) E741:ambiguous variable name 'l'
- (many instances) F541:f-string is missing placeholders
- (one or few instances) E721:do not compare types, use 'isinstance()'

Here the output:

```
2020-05-13T05:37:20.8820311Z ##[error]./pandas/_testing.py:630:22:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8824860Z ##[error]./pandas/_testing.py:1006:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8825837Z ##[error]./pandas/core/indexing.py:1569:42:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8826437Z ##[error]./pandas/core/series.py:2588:17:F541:f-string is missing placeholders
2020-05-13T05:37:20.8826992Z ##[error]./pandas/core/common.py:40:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8827546Z ##[error]./pandas/core/common.py:280:20:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8828505Z ##[error]./pandas/core/common.py:288:24:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8829122Z ##[error]./pandas/core/frame.py:685:37:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8830045Z ##[error]./pandas/core/indexes/multi.py:992:49:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8830652Z ##[error]./pandas/core/indexes/multi.py:1143:15:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8831243Z ##[error]./pandas/core/indexes/multi.py:1146:29:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8831817Z ##[error]./pandas/core/reshape/reshape.py:143:37:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8832381Z ##[error]./pandas/core/dtypes/concat.py:27:21:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8832963Z ##[error]./pandas/core/internals/concat.py:446:28:E721:do not compare types, use 'isinstance()'
2020-05-13T05:37:20.8833301Z ##[error]./pandas/core/arrays/interval.py:1054:5:E301:expected 1 blank line, found 0
2020-05-13T05:37:20.8833868Z ##[error]./pandas/util/_doctools.py:37:52:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8834428Z ##[error]./pandas/util/_doctools.py:38:48:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8834986Z ##[error]./pandas/util/_doctools.py:40:49:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8835554Z ##[error]./pandas/util/_doctools.py:41:49:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8836113Z ##[error]./pandas/util/_doctools.py:61:35:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8837945Z ##[error]./pandas/util/_doctools.py:76:55:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8838579Z ##[error]./pandas/util/_doctools.py:77:55:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8839206Z ##[error]./pandas/util/_doctools.py:91:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8839782Z ##[error]./pandas/plotting/_matplotlib/tools.py:374:9:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8840359Z ##[error]./pandas/plotting/_matplotlib/boxplot.py:147:29:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8840933Z ##[error]./pandas/plotting/_matplotlib/boxplot.py:148:43:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8842322Z ##[error]./pandas/plotting/_matplotlib/core.py:1508:56:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8843347Z ##[error]./pandas/io/pytables.py:4587:59:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8844029Z ##[error]./pandas/io/parsers.py:2971:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8845476Z ##[error]./pandas/io/parsers.py:3000:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8846243Z ##[error]./pandas/io/parsers.py:3020:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8847558Z ##[error]./pandas/io/formats/format.py:979:57:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8848615Z ##[error]./pandas/tests/groupby/test_categorical.py:505:13:F541:f-string is missing placeholders
2020-05-13T05:37:20.8849732Z ##[error]./pandas/tests/groupby/transform/test_numba.py:20:46:F541:f-string is missing placeholders
2020-05-13T05:37:20.8850374Z ##[error]./pandas/tests/groupby/transform/test_numba.py:23:46:F541:f-string is missing placeholders
2020-05-13T05:37:20.8851580Z ##[error]./pandas/tests/groupby/aggregate/test_numba.py:21:46:F541:f-string is missing placeholders
2020-05-13T05:37:20.8852714Z ##[error]./pandas/tests/groupby/aggregate/test_numba.py:24:46:F541:f-string is missing placeholders
2020-05-13T05:37:20.8853337Z ##[error]./pandas/tests/indexes/interval/test_interval.py:56:43:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8853951Z ##[error]./pandas/tests/indexes/interval/test_interval.py:77:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8854548Z ##[error]./pandas/tests/indexes/interval/test_constructors.py:333:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8855131Z ##[error]./pandas/tests/indexes/multi/test_indexing.py:797:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8855712Z ##[error]./pandas/tests/plotting/test_datetimelike.py:760:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8856293Z ##[error]./pandas/tests/plotting/test_datetimelike.py:787:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8856877Z ##[error]./pandas/tests/plotting/test_datetimelike.py:800:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8857459Z ##[error]./pandas/tests/plotting/test_datetimelike.py:866:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8858037Z ##[error]./pandas/tests/plotting/test_datetimelike.py:884:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8858611Z ##[error]./pandas/tests/plotting/test_datetimelike.py:997:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8859187Z ##[error]./pandas/tests/plotting/test_datetimelike.py:1006:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8859765Z ##[error]./pandas/tests/plotting/test_datetimelike.py:1131:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8860363Z ##[error]./pandas/tests/plotting/test_datetimelike.py:1135:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8860937Z ##[error]./pandas/tests/plotting/test_datetimelike.py:1236:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8861515Z ##[error]./pandas/tests/frame/methods/test_to_dict.py:121:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8862087Z ##[error]./pandas/tests/reshape/test_pivot.py:2069:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8862677Z ##[error]./pandas/tests/reshape/merge/test_merge.py:2183:38:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8863272Z ##[error]./pandas/tests/reshape/merge/test_merge.py:2183:69:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8863847Z ##[error]./pandas/tests/io/test_stata.py:1864:35:F541:f-string is missing placeholders
2020-05-13T05:37:20.8864409Z ##[error]./pandas/tests/io/test_sql.py:2402:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8864972Z ##[error]./pandas/tests/io/test_sql.py:2683:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8865545Z ##[error]./pandas/tests/io/formats/test_format.py:1801:31:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8866121Z ##[error]./pandas/tests/extension/test_interval.py:30:32:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8866713Z ##[error]./pandas/tests/indexing/test_floats.py:287:50:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8867287Z ##[error]./pandas/tests/indexing/test_floats.py:346:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8867867Z ##[error]./pandas/tests/indexing/test_floats.py:360:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8868581Z ##[error]./pandas/tests/indexing/test_floats.py:383:17:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8869229Z ##[error]./pandas/tests/indexing/test_floats.py:407:48:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8869822Z ##[error]./pandas/tests/indexing/test_floats.py:439:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8870388Z ##[error]./pandas/tests/indexing/test_floats.py:455:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8870956Z ##[error]./pandas/tests/indexing/test_floats.py:470:13:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8871524Z ##[error]./pandas/tests/indexing/test_floats.py:492:66:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8872091Z ##[error]./pandas/tests/indexing/test_floats.py:519:32:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8872862Z ##[error]./pandas/tests/indexing/test_loc.py:802:22:E741:ambiguous variable name 'l'
2020-05-13T05:37:20.8873460Z ##[error]./asv_bench/benchmarks/reshape.py:106:36:E741:ambiguous variable name 'l'
```

There is an additional issue with flake8-rst, e.g. here:

- https://gitlab.com/pycqa/flake8/-/issues/641
- https://github.com/kataev/flake8-rst/issues/22

Happy to have a go, but would like suggestions about how to proceed?  Suggestion:

Staged approach:
- [x] Pin flake < 3.8.0 to deal with flake8-rst issue
- [x] Ignore errors to get build green
- [x] Address F541
- [x] Address E721
- [ ] Address E741 (or not?)

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #34150 (maybe?)
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -43,3 +43,7 @@ repos:
         entry: python -m scripts.generate_pip_deps_from_conda
         files: ^(environment.yml|requirements-dev.txt)$
         pass_filenames: false
+-   repo: https://github.com/MarcoGorelli/flake8-rst-mirror
+    rev: v0.7.0
+    hooks:
+    -   id: flake8-rst
```
Comment (dsaxton) [inline .pre-commit-config.yaml:49]: Will this use the settings from setup.cfg?

Comment (MarcoGorelli) [reply .pre-commit-config.yaml:49]: @dsaxton yup - if I apply this
```diff
diff --git a/setup.cfg b/setup.cfg
index d938d2ef3..8b26424e4 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -34,7 +34,6 @@ exclude =
 
 [flake8-rst]
 bootstrap =
-    import numpy as np
     import pandas as pd
     np  # avoiding error when importing again numpy or pandas
     pd  # (in some cases we want to do it to show users)
```
then the hook will fail

File: requirements-dev.txt [hunk 2]
```diff
@@ -8,9 +8,8 @@ asv
 cython>=0.29.21
 black==20.8b1
 cpplint
-flake8<3.8.0
+flake8
```
Comment (MarcoGorelli) [inline requirements-dev.txt:11]: @erfannariman I only modified `environment.yml` and then the hook you added modified this file for me :clap:

Comment (erfannariman) [reply requirements-dev.txt:11]: That's pretty cool! Saves quite some CI errors for other devs in the future.

File: environment.yml [hunk 3]
```diff
@@ -17,9 +17,8 @@ dependencies:
   # code checks
   - black=20.8b1
   - cpplint
-  - flake8<3.8.0  # temporary pin, GH#34150
+  - flake8
```
Comment (simonjayhawkins) [inline environment.yml:20]: you can probably remove the pip install of pyflakes>=2.2.0 now.

**Comment by MarcoGorelli** [comment]
Turns out this can be simplified - no need for a fork/mirror, we can just pin `flake8-rst` and `flake8` in `additional_dependencies` and run it as a local hook

**Comment by fangchenli** [comment]
The new pip dependency resolver will become default soon https://discuss.python.org/t/announcement-pip-20-2-release/4863. I tested it earlier. It got stuck if `flake8` is pinned using `<` sign. 

Also
```
The conflict is caused by:
    The user requested pyflakes>=2.2.0
    flake8 3.7.9 depends on pyflakes<2.2.0 and >=2.1.0
```

So we probibly need to merge this before pip 20.3 release.

**Comment by jreback** [comment]
can you merge master and ping on green.

Is there a way to have an option in code_checks.sh that actually runs the appropriate pre-commit hook? that way we have a single source of truth, and you can actually still run them (of course have to *avoid* running these on the CI from the script as well).

**Comment by MarcoGorelli** [comment]
Sure, done

> Is there a way to have an option in code_checks.sh that actually runs the appropriate pre-commit hook? 

Have opened #36956 for this

**Comment by jreback** [comment]
lgtm. can you merge master. merge on green (just to be sure nothing broken since this).

**Comment by jreback** [comment]
thanks @MarcoGorelli

**Pull request merged**